src/codegen/compile.ml

```diff
@@ -4164,7 +4164,81 @@ module RTS_Exports = struct
 
 end (* RTS_Exports *)
 
-module Serialization = struct
+module type Stream = sig
+  (*               env    ref    code *)
+  val write_byte : E.t -> G.t -> G.t -> G.t
+  val write_word_leb : E.t -> G.t -> G.t -> G.t
+  val write_word_32 : E.t -> G.t -> G.t -> G.t
+  val write_blob : E.t -> G.t -> G.t -> G.t
+  val write_text : E.t -> G.t -> G.t -> G.t
+  val write_unsigned : E.t -> G.t -> G.t -> G.t
+  val write_signed : E.t -> G.t -> G.t -> G.t
+
+  (* opportunity to flush or update, stream token is on stack *)
+  val checkpoint : E.t -> G.t -> G.t
+
+  (* reserve a small fixed number of bytes in the stream and return an address to it *)
+  val reserve : E.t -> G.t -> int32 -> G.t
+end
+
+
+module BumpStream : Stream = struct
+  let advance_data_buf get_data_buf =
+    get_data_buf ^^ G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^ G.setter_for get_data_buf
+
+  let checkpoint _env get_data_buf = G.setter_for get_data_buf
```

Comment:
Haven't seen `setter_for` yet … hmm, feels a bit hacky to me. Maybe really worth introducing a type to abstract over things that can be set and get'ed.

Reply:
I believe you reviewed my "hack" back then :-)

I don't see how extracting the local index from the getter is a bad thing. A Van Laarhoven lens-like approach would be ideal, but will OCaml's type system carry the weight? I still want `get_data_buf : G.t` in all contexts!

Reply:
I'd redactor this to `get data_buf : G.t` where `data_buf` is of a data type that's either a local or global. But not in this PR

Reply:
And I believe I tried that before, so this may hit some snag somewhere else

Reply:
I kinda agree with Joachim. Also, get_data_buf below is not a local but a function parameter. I know wasm function parameters are index as locals, but are they actually updateable? the implementation of share_code makes them so by declaring extra locals, but that's a bit gross.

Reply:
Yes, they are updateable, we use this in other places as well